[CRITICAL] Orphaned Conditional Algo Orders: cancel_oco_pair uses wrong Binance API, exhausts 100-order cap and blocks all trading

# 🤖 BMAD Incident Resolution Ticket: INC-2026-04-15-orphaned-algo-orders

## 🚨 Status: READY FOR AUTONOMOUS EXECUTION
**Priority**: P0 — Critical (system fully blocked from trading)
**Source**: Manual investigation (2026-04-15T17:19:00Z)

---

## 🔍 Investigation Artifacts
- **Evidence Snapshot**: `petrosa_k8s/_bmad-output/incidents/2026-04-15/evidence.json`
- **Reproduction Test**: `petrosa_k8s/_bmad-output/incidents/2026-04-15/reproduction_test.py`
- **Hypothesis Confidence**: 98/100

---

## 🧩 Root Cause Hypothesis

**Description**: `cancel_oco_pair()` calls `futures_cancel_batch_orders()` (standard Binance Futures REST API) to cancel SL/TP orders. However, SL and TP are placed as **Conditional Algo Orders** via `POST /fapi/v1/algo/algoOrder`. The standard batch-cancel endpoint (`DELETE /fapi/v1/batchOrders`) does not cancel algo orders — it silently succeeds or returns an empty result. The only correct cancel path is `DELETE /fapi/v1/algo/algoOrder` with `algoId`. A fallback to this endpoint exists in `cancel_order()` (triggered by error `-2011`/`-4132`), but `cancel_oco_pair()` uses `futures_cancel_batch_orders` which never raises those codes.

Over time, every position close leaves 2 orphaned conditional algo orders on Binance. These accumulate until the 100-order account cap is reached, at which point all new OCO order placements are rejected and signal execution is fully blocked.

**Blast Radius**:
- All trade symbols blocked simultaneously once cap is hit (BTCUSDT, ETHUSDT, BNBUSDT, XRPUSDT, DOTUSDT, LTCUSDT, BCHUSDT, XLMUSDT, ADAUSDT, LINKUSDT)
- 96 orphaned orders confirmed on 2026-04-15 (96/100 cap used)
- System was 100% unable to execute any new signals

**Suspected Code Location**:
- File: `tradeengine/dispatcher.py`
- Symbol: `OCOManager.cancel_oco_pair`
- Lines: `274–346`

---

## 🧪 Mandatory Reproduction Test (FAILED)

The following test was generated and executed. It **FAILS** with the exact incident signature. You must fix the code until this test passes.

```python
# File: petrosa_k8s/_bmad-output/incidents/2026-04-15/reproduction_test.py
# Test: TestCancelOcoPairUsesWrongAPI::test_cancel_oco_pair_silently_fails_for_algo_orders

from tradeengine.dispatcher import OCOManager
import logging
from unittest.mock import MagicMock

mock_exchange = MagicMock()
mock_exchange.client.futures_cancel_batch_orders.return_value = []
algo_cancel_calls = []
mock_exchange.client._request_futures_api = MagicMock(side_effect=lambda *a, **kw: algo_cancel_calls.append(kw))

oco_manager = OCOManager(exchange=mock_exchange, logger=logging.getLogger("test"))
oco_manager.active_oco_pairs["BTCUSDT_LONG"] = [{
    "symbol": "BTCUSDT", "position_side": "LONG",
    "sl_order_id": 1000000049928499, "tp_order_id": 1000000049928500,
    "status": "active", "reconciled": False,
    "position_id": "test-001", "strategy_position_id": "strat-001",
    "exchange_position_key": "BTCUSDT_LONG",
}]

await oco_manager.cancel_oco_pair(position_id="test-001", symbol="BTCUSDT", position_side="LONG")
assert len(algo_cancel_calls) == 2, "cancel_oco_pair must call algo-order DELETE API for SL and TP"
```

**Traceback Captured**:
```text
FAILED reproduction_test.py::TestCancelOcoPairUsesWrongAPI::test_cancel_oco_pair_silently_fails_for_algo_orders

tradeengine/dispatcher.py, line ~318, in cancel_oco_pair:
    cancel_result = self.exchange.client.futures_cancel_batch_orders(
        symbol=oco_symbol, orderIdList=[sl_order_id, tp_order_id]
    )
    # Returns [] silently — algo orders are NOT cancelled
    # No BinanceAPIException raised → no fallback to algo API triggered
    # oco_info["status"] set to "cancelled" in memory — but orders remain on Binance

AssertionError: cancel_oco_pair must call algo-order DELETE API for SL and TP
# algo_cancel_calls == 0 (should be 2)
```

**Live system evidence** (from kubectl logs, 2026-04-15):
```
⛔ RISK REJECTION: Algo order limit reached for XRPUSDT (11/10 orders). Cannot place OCO.
⛔ RISK REJECTION: Algo order limit reached for BCHUSDT (9/10 orders). Cannot place OCO.
Retrieved 0 open positions via Data Manager
[repeated for every incoming signal — system fully blocked]
```

---

## 🛠️ Suggested Remediation

In `tradeengine/dispatcher.py`, `OCOManager.cancel_oco_pair()` (~line 318), replace `futures_cancel_batch_orders` with individual `DELETE /fapi/v1/algo/algoOrder` calls:

```python
# REMOVE:
cancel_result = self.exchange.client.futures_cancel_batch_orders(
    symbol=oco_symbol, orderIdList=[sl_order_id, tp_order_id]
)
if cancel_result and len(cancel_result) >= 2:
    oco_info["status"] = "cancelled"
else:
    all_cancelled = False

# REPLACE WITH:
for order_id, label in [(sl_order_id, "SL"), (tp_order_id, "TP")]:
    try:
        self.exchange.client._request_futures_api(
            "delete", "algoOrder", signed=True,
            data={"symbol": oco_symbol, "algoId": order_id},
        )
        self.logger.info(f"✅ Cancelled algo {label} order {order_id} for {oco_symbol}")
    except Exception as e:
        self.logger.error(f"❌ Failed to cancel algo {label} {order_id}: {e}")
        all_cancelled = False
oco_info["status"] = "cancelled"
```

---

## 📋 Acceptance Criteria (for `ticket-execution`)

- [ ] Reproducing test `reproduction_test.py::TestCancelOcoPairUsesWrongAPI` passes 100%.
- [ ] No regression in related tests in `tests/test_oco_orders.py` and `tests/integration/test_oco_fill_cancel.py`.
- [ ] After a position closes, corresponding SL and TP algo orders are absent from `GET /fapi/v1/openAlgoOrders` within 5 seconds. Verified by integration test or manual check.
- [ ] Fix adheres to project conventions and ADR-001 principles.

--- a/tests/test_oco_orders.py
+++ b/tests/test_oco_orders.py
@@ -56,11 +56,14 @@ def mock_exchange():
 
     exchange.execute = AsyncMock(side_effect=mock_execute)
 
-    # Mock batch order cancellation
-    def mock_cancel_batch(symbol: str, orderIdList: list) -> list:
-        return [{"orderId": oid, "status": "CANCELED"} for oid in orderIdList]
+    # Mock algo order cancellation (DELETE /fapi/v1/algo/algoOrder)
+    def mock_request_futures_api(method, path, signed=False, data=None, **kwargs):
+        if method == "delete" and path == "algoOrder":
+            algo_id = (data or {}).get("algoId")
+            return {"algoId": algo_id, "algoStatus": "CANCELLED"}
+        return {}
 
-    exchange.client.futures_cancel_batch_orders = mock_cancel_batch
+    exchange.client._request_futures_api = Mock(side_effect=mock_request_futures_api)
 
     # Mock single order cancellation
     def mock_cancel_order(symbol: str, orderId: str) -> dict:
@@ -235,7 +238,7 @@ async def test_place_oco_orders_short_position(oco_manager: OCOManager):
 
 
 @pytest.mark.asyncio
-async def test_cancel_oco_pair(oco_manager: OCOManager):
+async def test_cancel_oco_pair(oco_manager: OCOManager, mock_exchange):
     """Test cancelling both SL and TP orders"""
     # First, place OCO orders
     await oco_manager.place_oco_orders(
@@ -261,6 +264,12 @@ async def test_cancel_oco_pair(oco_manager: OCOManager):
         oco_list = oco_manager.active_oco_pairs[exchange_key]
         if len(oco_list) > 0:
             assert oco_list[0]["status"] == "cancelled"
+
+    # Verify algo cancel API was called (not the wrong batch cancel)
+    mock_exchange.client._request_futures_api.assert_called()
+    assert not mock_exchange.client.futures_cancel_batch_orders.called, (
+        "futures_cancel_batch_orders must NOT be called for algo orders"
+    )
 
     # Clean up
     await oco_manager.stop_monitoring()

--- a/tests/test_oco_orders.py
+++ b/tests/test_oco_orders.py
@@ -240,8 +240,8 @@ async def test_place_oco_orders_short_position(oco_manager: OCOManager):
 @pytest.mark.asyncio
 async def test_cancel_oco_pair(oco_manager: OCOManager, mock_exchange):
     """Test cancelling both SL and TP orders"""
-    # First, place OCO orders
-    await oco_manager.place_oco_orders(
+    # First, place OCO orders and capture the returned IDs
+    place_result = await oco_manager.place_oco_orders(
         position_id="test_pos_cancel_789",
         symbol="BTCUSDT",
         position_side="LONG",
@@ -249,6 +249,11 @@ async def test_cancel_oco_pair(oco_manager: OCOManager, mock_exchange):
         stop_loss_price=48000.0,
         take_profit_price=52000.0,
     )
+    sl_id = place_result["sl_order_id"]
+    tp_id = place_result["tp_order_id"]
+
+    # Reset call history so we only track cancellation calls
+    mock_exchange.client._request_futures_api.reset_mock()
 
     # Cancel the OCO pair (need to pass symbol and position_side for new key structure)
     result = await oco_manager.cancel_oco_pair(
@@ -265,8 +270,31 @@ async def test_cancel_oco_pair(oco_manager: OCOManager, mock_exchange):
         if len(oco_list) > 0:
             assert oco_list[0]["status"] == "cancelled"
 
-    # Verify algo cancel API was called (not the wrong batch cancel)
-    mock_exchange.client._request_futures_api.assert_called()
+    # Verify algo cancel API called exactly twice (SL + TP) with correct algoIds
+    from unittest.mock import call
+
+    assert mock_exchange.client._request_futures_api.call_count == 2, (
+        "Must call algo DELETE API exactly once for SL and once for TP"
+    )
+    mock_exchange.client._request_futures_api.assert_has_calls(
+        [
+            call(
+                "delete",
+                "algoOrder",
+                signed=True,
+                data={"symbol": "BTCUSDT", "algoId": sl_id},
+            ),
+            call(
+                "delete",
+                "algoOrder",
+                signed=True,
+                data={"symbol": "BTCUSDT", "algoId": tp_id},
+            ),
+        ],
+        any_order=False,
+    )
+
+    # Verify batch cancel API was NOT used
     assert not mock_exchange.client.futures_cancel_batch_orders.called, (
         "futures_cancel_batch_orders must NOT be called for algo orders"
     )

--- a/tradeengine/dispatcher.py
+++ b/tradeengine/dispatcher.py
@@ -326,6 +326,14 @@ class OCOManager:
 
             pair_cancelled = True
             for order_id, label in [(sl_order_id, "SL"), (tp_order_id, "TP")]:
+                if not order_id:
+                    self.logger.warning(
+                        f"⚠️ Missing {label} order ID for {oco_symbol} — "
+                        f"skipping cancel, pair may need reconciliation"
+                    )
+                    pair_cancelled = False
+                    all_cancelled = False
+                    continue
                 try:
                     self.exchange.client._request_futures_api(
                         "delete",

--- a/tradeengine/dispatcher.py
+++ b/tradeengine/dispatcher.py
@@ -324,24 +324,29 @@ class OCOManager:
                 f"sl_order_id={sl_order_id}, tp_order_id={tp_order_id}"
             )
 
-            try:
-                # Cancel both orders using batch cancellation
-                cancel_result = self.exchange.client.futures_cancel_batch_orders(
-                    symbol=oco_symbol, orderIdList=[sl_order_id, tp_order_id]
-                )
-
-                if cancel_result and len(cancel_result) >= 2:
-                    self.logger.info(
-                        f"✅ OCO PAIR CANCELLED for strategy {oco_info.get('strategy_position_id')}"
+            pair_cancelled = True
+            for order_id, label in [(sl_order_id, "SL"), (tp_order_id, "TP")]:
+                try:
+                    self.exchange.client._request_futures_api(
+                        "delete",
+                        "algoOrder",
+                        signed=True,
+                        data={"symbol": oco_symbol, "algoId": order_id},
                     )
-                    oco_info["status"] = "cancelled"
-                else:
-                    self.logger.error(f"❌ FAILED TO CANCEL OCO PAIR: {cancel_result}")
+                    self.logger.info(
+                        f"✅ Cancelled algo {label} order {order_id} for {oco_symbol}"
+                    )
+                except Exception as e:
+                    self.logger.error(
+                        f"❌ Failed to cancel algo {label} order {order_id}: {e}"
+                    )
+                    pair_cancelled = False
                     all_cancelled = False
-
-            except Exception as e:
-                self.logger.error(f"❌ ERROR CANCELLING OCO PAIR: {e}")
-                all_cancelled = False
+            if pair_cancelled:
+                self.logger.info(
+                    f"✅ OCO PAIR CANCELLED for strategy {oco_info.get('strategy_position_id')}"
+                )
+                oco_info["status"] = "cancelled"
 
         return all_cancelled
 

